ldpd/lde.c

```diff
@@ -1202,6 +1216,89 @@ lde_send_labelrelease(struct lde_nbr *ln, struct fec_node *fn,
 	lde_imsg_compose_ldpe(IMSG_RELEASE_ADD_END, ln->peerid, 0, NULL, 0);
 }
 
+void
+lde_send_labelrequest(struct lde_nbr *ln, struct fec_node *fn,
+		      struct map *wcard, int single)
+{
+	struct map		 map;
+	struct fec		*f;
+	struct lde_req		*lre;
+
+	if (fn) {
+		lde_fec2map(&fn->fec, &map);
+		switch (fn->fec.type) {
+		case FEC_TYPE_IPV4:
+			if (!ln->v4_enabled)
+				return;
+			if (lde_acl_check(ldeconf->ipv4.acl_label_advertise_to,
+				AF_INET, (union ldpd_addr *)&ln->id, 32)
+				!= FILTER_PERMIT)
+				return;
+			if (lde_acl_check(ldeconf->ipv4.acl_label_advertise_for,
+				AF_INET,
+				(union ldpd_addr *)&fn->fec.u.ipv4.prefix,
+				fn->fec.u.ipv4.prefixlen) != FILTER_PERMIT)
+				return;
+			break;
+		case FEC_TYPE_IPV6:
+			if (!ln->v6_enabled)
+				return;
+			if (lde_acl_check(ldeconf->ipv6.acl_label_advertise_to,
+				AF_INET, (union ldpd_addr *)&ln->id, 32)
+				!= FILTER_PERMIT)
+				return;
+			if (lde_acl_check(ldeconf->ipv6.acl_label_advertise_for,
+				AF_INET6,
+				(union ldpd_addr *)&fn->fec.u.ipv6.prefix,
+				fn->fec.u.ipv6.prefixlen) != FILTER_PERMIT)
+				return;
+			break;
+		default:
+			fatalx("lde_send_labelrequest: unknown af");
+		}
+	} else
+		memcpy(&map, wcard, sizeof(map));
+
+	map.label = NO_LABEL;
+
+	/* SLRq.3: send label request */
+	lde_imsg_compose_ldpe(IMSG_REQUEST_ADD, ln->peerid, 0, &map,
+	      sizeof(map));
+
+	if (single)
+		lde_imsg_compose_ldpe(IMSG_REQUEST_ADD_END,
+		       ln->peerid, 0, NULL, 0);
+
+	/* SLRq.4: record sent request */
+	if (fn) {
+		lre = (struct  lde_req *)fec_find(&ln->sent_req, &fn->fec);
+		if (lre == NULL)
+			lde_req_add(ln, &fn->fec, 1);
+	} else {
+		RB_FOREACH(f, fec_tree, &ft) {
+			fn = (struct fec_node *)f;
+			lre = (struct lde_req *)fec_find(&ln->sent_req, &fn->fec);
+			if (lde_wildcard_apply(wcard, &fn->fec, NULL) == 0)
+				continue;
+			if (lre == NULL)
+				lde_req_add(ln, f, 1);
+		}
+	}
+
+}
+
+void
+lde_send_labelrequest_wcard(struct lde_nbr *ln, uint16_t af)
+{
+	struct map	 wcard;
+
+	memset(&wcard, 0, sizeof(wcard));
+	wcard.type = MAP_TYPE_TYPED_WCARD;
```

Comment:
I though from reading the RFC 5036 that wildcard could only be used for labelwithdraw and labelrelease.    
In RFC 5918  labelrequest wild card was added and is only supported with this typed wildcard.   But I'm probably missing something here in the RFC

Reply:
You're right. It this case it seems the only option is to reset the neighborships as we discussed before.

Reply:
I'm assuming the correct behavior here is for lde process to send a message to ldpe and then in ldpe call 
session_shutdown(nbr, S_SHUTDOWN, 0, 0);  This will result in a IMSG_NEIGHBOR_DOWN to lde

From RFC 5036 section  3.9.  Status Code Summary I could not find an error that indicated to restart.

Reply:
I think `S_SHUTDOWN` is the right status code for that.

Reply:
fixed